### PR TITLE
Introduce a fixed notion of time for the duration of the update cycle

### PR DESCRIPTION
**Pull request by joshuagl**
Addresses: #113 

Per the discussion between @erickt and @trishankatdatadog in Slack, this PR introduces a fixed notion of time to the client workflow. The client should store the time at the beginning of the update and use it for comparing expiration times throughout the detailed client workflow, so long as the session has not lasted longer than the re-signing period of the timestamp metadata.

### CODE REVIEW

File: tuf-spec.md
Comment (JustinCappos) [reply tuf-spec.md:1141]: Also, I wonder if we need to keep repeating the suggestion about setting the update workflow maximum duration.  Might we be better off just saying this once?

File: tuf-spec.md
Comment (JustinCappos) [inline tuf-spec.md:1082]:
```suggestion
  beginning of the update workflow to prevent metadata from expiring during an
```

**Review by JustinCappos — COMMENTED** [review]
I want to discuss / think more about this.  I agree with the sentiment that we need to be more precise about how to handle time, but I want to make sure we're not opening up a can of worms in the slow retrieval case.

File: tuf-spec.md
Comment (erickt) [reply tuf-spec.md:32]: I'm just about to submit a typo-cleanup PR. I'll include these changes there.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1141]: I wasn't entirely comfortable with this being the timestamp re-signing period. At first it seemed like a reasonable time period that was already defined by the system. However, it's not a period we define or discuss explicitly in the specification. 

Perhaps it makes more sense to have the update workflow maximum duration be explicitly application defined? I'd like it if we suggested a sane default value, but it might be more abstract. Like when we talk about downloading the root metadata and suggest

> up to some W number of bytes (because the size is unknown). The value for W is set by the authors of the application using TUF. For example, W may be tens of kilobytes.

maybe something like

> The latest known time should be lower than the update start time plus the update workflow maximum duration, D.  The value for D is set by the authors of the application using TUF. For example, D may be tens of minutes.

Agreed that it would be better to define the update workflow maximum duration in one place, we can do that at the same time as we record the beginning of the update in step 0. I will make that change.

**Comment by lukpueh** [comment]
I cut a bunch of patch releases today that conflict with this PR. The pending #119 will create further conflicts. Give that I already cleaned up #119 so that it now merges cleanly with `master`, I suggest to first land #119 and then this PR. I am happy to then help resolving conflicts.

**Comment by joshuagl** [comment]
Updated the PR to only introduce the fixed time changes, to remove redundant re-definition of the fixed update window and rebase on the recent changes.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md]: I've removed the explicit suggestion of using the signing period of the timestamp metadata and instead suggested that the time period is application defined.

**Comment by joshuagl** [comment]
> I want to discuss / think more about this. I agree with the sentiment that we need to be more precise about how to handle time, but I want to make sure we're not opening up a can of worms in the slow retrieval case.

Hi @JustinCappos – did you have any further thoughts on this? Would you me to schedule a discussion to explore your concerns?

**Comment by davidstrauss** [comment]
Coming from the side of PHP-TUF, we've gone with a library design that receives a timestamp for "now" when instantiated. This not only meets the goal of a consistent time for all sequential operations within the library but also allows better testing integration (e.g. with static fixtures) and framework integration (e.g. with ones like Drupal that already capture and retain a sense of "now" during each run that we can also use in TUF).

Any outer layer wrapping the library, like a CLI, unit test, or daemon, would then have responsibility for providing "now" to the library.

**Comment by JustinCappos** [comment]
> > I want to discuss / think more about this. I agree with the sentiment that we need to be more precise about how to handle time, but I want to make sure we're not opening up a can of worms in the slow retrieval case.
> 
> Hi @JustinCappos – did you have any further thoughts on this? Would you me to schedule a discussion to explore your concerns?

I think I'm okay with this now.  In general, the concept of tracking the time at the beginning, having a window of time an update must complete in, etc. makes a lot of sense to me.

**Review by trishankatdatadog — APPROVED** [review]

**Comment by trishankatdatadog** [comment]
Am I the only one who finds this language unintuitive on first read?

`The fixed update expiration time MUST be lower than the expiration timestamp in the current X metadata file.`

Would it be helpful to paraphrase it to something more intuitive? For example:

`The expiration timestamp in the current X metadata file MUST be higher than the current fixed notion of now.`

Cc @jhdalek55

**Comment by joshuagl** [comment]
> Am I the only one who finds this language unintuitive on first read?

I suspect not. I tried to maintain the original phrasing of the freeze attack sections, but that was probably not necessary.

I couldn't think of a reason that "fixed now < expiration" was clearer than "expiration > fixed now", so I made the change you suggested. Thanks!

**Review by trishankatdatadog — DISMISSED** [review]

**Comment by joshuagl** [comment]
In the spirit of #132, could I get a second review/approval from @JustinCappos, @mnm678 or @lukpueh?

File: tuf-spec.md
Comment (mnm678) [inline tuf-spec.md:1277]:
```suggestion
  any metadata requested in steps 5.5.6.1 - 5.5.6.2 cannot be downloaded nor
```

**Review by mnm678 — DISMISSED** [review]

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md]: Well spotted, thank you! I've squashed that change into c2c4308

**Comment by joshuagl** [comment]
I forced push to squash in some of the fixup commits introduced during code review, and that has dismissed the existing reviews.

Will need re-approval to meet the branch protection rules. Thank you in advance.

**Review by mnm678 — APPROVED** [review]

**Comment by lukpueh** [comment]
@joshuagl, would you mind removing the merge commit form the PR?

**Comment by joshuagl** [comment]
> @joshuagl, would you mind removing the merge commit form the PR?

Sure, done!

(I was trying to figure out whether I could avoid it as force pushes now dismiss reviews. Alas)

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1118]: There are two more "On the next update cycle, begin at step 5.0" below. Were they not changed for a reason?

File: tuf-spec.md
Comment (lukpueh) [inline tuf-spec.md:1297]:
```suggestion
      then jump to step 5.6.
```

**Review by lukpueh — CHANGES_REQUESTED** [review]
The actual additions look good to me...

**Comment by lukpueh** [comment]
I'm looking forward to #121, when we no longer have to manually keep track of section numbers.

File: tuf-spec.md
Comment (joshuagl) [reply tuf-spec.md:1118]: They were not. I thought I'd caught all of the cross-references. 🤦

**Comment by lukpueh** [comment]
Btw. in [Section 7.2. we also reference "5.1"]( https://github.com/theupdateframework/specification/blame/890b156939116aaf4a0ffa7260a63be1a0c501a1/tuf-spec.md#L1437).

Judging from the git history, it means to [reference the section "The client application"](https://github.com/theupdateframework/specification/commit/daf0e4999abf3076fdcbed5648f09071d65b7cb8#diff-b0020aab85257910a5a31f248c201a81cb35e1939545c4b45515e9049e78cdffR979), which has no longer a numbering. Should we quickly fix that too?

**Comment by joshuagl** [comment]
> reference the section "The client application"

Well spotted 👀 . I was going to make the reference link to "The client application" but decided against, because none of the other section references are hyperlinks and should will be easier to implement (and maintain) after #121

**Review by lukpueh — APPROVED** [review]
Thanks!! ❤️

**Comment by lukpueh** [comment]
@mnm678, could you please re-approve?

**Review by mnm678 — APPROVED** [review]

**Comment by trishankatdatadog** [comment]
For anyone interested, I tested [[1](https://github.com/trishankatdatadog/tuf-on-a-plane/blob/0f51b6e10046e92726c893a5d79201e91ec897ec/src/tuf_on_a_plane/config.py#L40), [2](https://github.com/trishankatdatadog/tuf-on-a-plane/blob/0f51b6e10046e92726c893a5d79201e91ec897ec/src/tuf_on_a_plane/repository.py#L51)] that this idea works 🙂

**Pull request merged**
